R/split_data_by_task.R

```diff
@@ -0,0 +1,48 @@
+#' Split the input data by a single task.
+#'
+#' @description This function splits the input data by a single task with a
+#' combination of horizon, location, and target_end_date
+#'
+#' @param valid_tbl A data.frame containing forecast data and target data,
+#' processed by the function `validate_input_data()`.
+#'
+#' @return A list of data sets, each corresponding to a single task.
+#' @export
+#'
+#' @examples \dontrun{
+#' library(dplyr)
+#' library(hubExamples)
+#' forecast_data <- hubExamples::forecast_outputs |>
+#'   dplyr::filter(
+#'     output_type %in% c("quantile"),
+#'     location == "25",
+#'     horizon == 1
+#'   )
+#' target_data <- hubExamples::forecast_target_ts |>
+#'   dplyr::filter(
+#'     date %in% unique(forecast_data$target_end_date),
+#'     location == "25"
+#'   ) |>
+#'   # Rename columns to match the oracle output format
+#'   rename(
+#'     target_end_date = date,
+#'     oracle_value = observation
+#'   )
+#' valid_tbl <- validate_input_data(forecast_data, target_data)
+#' split_data_by_task(valid_tbl)
+#' }
+split_data_by_task <- function(valid_tbl) {
+  # Get columns to use for splitting data by task
+  split_cols <- c("horizon", "location", "target_end_date")
+
+  # List of data sets split by task
+  datasets_by_task <- valid_tbl |>
+    dplyr::group_by(
```

Comment:
Could this nested set of dplyr calls be unnested and instead strung together with pipes? It seems like this nesting is exactly what the pipes are supposed to help make more legible. Like, could it be instead:
```
        dplyr::all_of(split_cols) |>
        dplyr::across() |>
       dplyr::group_by() |>
```

Reply:
Nesting is necessary because the `dplyr::across ()` and `dplyr::all_of()` functions must only be used inside data-masking verbs like `mutate()`, `filter()`, and `group_by()`. I removed `dplyr::` for them instead, as those functions are only called in a `dplyr` context.